File: sdk/ios/Sources/FraudShield/OtpGuard.swift
import Foundation
import CallKit
import UIKit

// MARK: - OtpGuard (iOS)
//
// Protects OTP screens from social engineering disclosure.
//
// INTEGRATION:
//   class OtpViewController: UIViewController {
//       private let otpGuard = OtpGuard.shared
//
//       override func viewWillAppear(_ animated: Bool) {
//           super.viewWillAppear(animated)
//           otpGuard.activate(on: self) { riskLevel in
//               if riskLevel == .highRisk {
//                   // Optionally auto-cancel the OTP request
//               }
//           }
//       }
//
//       override func viewWillDisappear(_ animated: Bool) {
//           super.viewWillDisappear(animated)
//           otpGuard.deactivate()
//       }
//   }

public enum OtpRiskLevel {
    case clean        // Not on a call
    case mediumRisk   // On a known call — warning shown
    case highRisk     // On a call with unknown caller — full block shown
}

public class OtpGuard: NSObject {
    public static let shared = OtpGuard()

    private let callObserver = CXCallObserver()
    private var warningViewController: UIViewController?
    private weak var hostController: UIViewController?

    private override init() {
        super.init()
        callObserver.setDelegate(self, queue: .main)
    }

    // MARK: - Public API

    /// Call when the OTP screen appears. Checks for active call and applies protections.
    public func activate(on viewController: UIViewController, completion: @escaping (OtpRiskLevel) -> Void) {
        hostController = viewController

        // Prevent screenshots — marks the window as secure
        viewController.view.window?.isHidden = false
        setScreenshotProtection(enabled: true, on: viewController)

        let activeCalls = callObserver.calls.filter { !$0.hasEnded }
        guard !activeCalls.isEmpty else {
            completion(.clean)
            return
        }

        let isConnected = activeCalls.contains { $0.hasConnected }
        guard isConnected else {
            completion(.clean)
            return
        }

        // We can't easily determine unknown caller on iOS without Contacts access.
        // Use CallStateCollector's contact-check logic instead.
        let callCollector = CallStateCollector()
        let riskLevel: OtpRiskLevel = callCollector.isCallerUnknown() ? .highRisk : .mediumRisk

        showWarningOverlay(on: viewController, riskLevel: riskLevel)
        fireOtpOnCallSignal(riskLevel: riskLevel)
        completion(riskLevel)
    }

    /// Call when OTP screen disappears.
    public func deactivate() {
        removeWarningOverlay()
        if let host = hostController {
            setScreenshotProtection(enabled: false, on: host)
        }
        hostController = nil
    }

    // MARK: - Screenshot Prevention

    private func setScreenshotProtection(enabled: Bool, on viewController: UIViewController) {
        // iOS does not provide a FLAG_SECURE equivalent, but we can obscure
        // the OTP field using a secure text field approach — the host app's
        // OTP fields should use UITextField.isSecureTextEntry = true.
        // We add an opaque UITextField overlay trick to prevent system screenshot capture.
        if enabled {
            let secureField = UITextField()
            secureField.isSecureTextEntry = true
            secureField.frame = viewController.view.bounds
            secureField.backgroundColor = .clear
            secureField.isUserInteractionEnabled = false
            secureField.tag = 9998
            viewController.view.addSubview(secureField)
            viewController.view.layer.superlayer?.addSublayer(secureField.layer)
            secureField.layer.sublayers?.last?.addSublayer(viewController.view.layer)
        } else {
            viewController.view.subviews.first(where: { $0.tag == 9998 })?.removeFromSuperview()
        }
    }

    // MARK: - Warning Overlay

    private func showWarningOverlay(on presenter: UIViewController, riskLevel: OtpRiskLevel) {
        let vc = OtpWarningViewController(riskLevel: riskLevel)
        vc.modalPresentationStyle = .overFullScreen
        vc.modalTransitionStyle = .crossDissolve
        warningViewController = vc
        presenter.present(vc, animated: true)
    }

    private func removeWarningOverlay() {
        warningViewController?.dismiss(animated: true)
        warningViewController = nil
    }

    // MARK: - Risk Signal

    private func fireOtpOnCallSignal(riskLevel: OtpRiskLevel) {
        Task {
            let payload: [String: Any] = [
                "event_type": "OTP_SCREEN_ON_CALL",
                "unknown_caller": riskLevel == .highRisk,
                "rule": "RULE_011",
                "risk_delta": riskLevel == .highRisk ? 80 : 45,
                "timestamp": Date().timeIntervalSince1970,
            ]
            try? await FraudSignalDispatcher.shared.dispatchEvent(payload)
        }
    }
}

// MARK: - CXCallObserverDelegate

extension OtpGuard: CXCallObserverDelegate {
    public func callObserver(_ callObserver: CXCallObserver, callChanged call: CXCall) {
        // If a call ends while OTP screen is still active, remove the overlay
        if call.hasEnded, warningViewController != nil {
            removeWarningOverlay()
        }
    }
}

// MARK: - Warning View Controller

private class OtpWarningViewController: UIViewController {
    private let riskLevel: OtpRiskLevel

    init(riskLevel: OtpRiskLevel) {
        self.riskLevel = riskLevel
        super.init(nibName: nil, bundle: nil)
    }

    required init?(coder: NSCoder) { fatalError() }

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = UIColor(red: 0.78, green: 0.10, blue: 0.10, alpha: 0.95)

        let stack = UIStackView()
        stack.axis = .vertical
        stack.spacing = 16
        stack.alignment = .center
        stack.translatesAutoresizingMaskIntoConstraints = false
        view.addSubview(stack)

        NSLayoutConstraint.activate([
            stack.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            stack.centerYAnchor.constraint(equalTo: view.centerYAnchor),
            stack.leadingAnchor.constraint(equalTo: view.leadingAnchor, constant: 32),
            stack.trailingAnchor.constraint(equalTo: view.trailingAnchor, constant: -32),
        ])

        let emoji = UILabel()
        emoji.text = riskLevel == .highRisk ? "🚨" : "⚠️"
        emoji.font = .systemFont(ofSize: 56)
        stack.addArrangedSubview(emoji)

        let headline = UILabel()
        headline.text = riskLevel == .highRisk ? "SCAM ALERT" : "Active Call Detected"
        headline.font = .boldSystemFont(ofSize: 22)
        headline.textColor = .white
        headline.textAlignment = .center
        stack.addArrangedSubview(headline)

        let body = UILabel()
        body.numberOfLines = 0
        body.textAlignment = .center
        body.textColor = UIColor(white: 0.9, alpha: 1)
        body.font = .systemFont(ofSize: 15)
        body.text = riskLevel == .highRisk
            ? "You are on a call with an UNKNOWN number.\n\n⛔ NEVER read your OTP to anyone.\n⛔ MTN will never ask for your code.\n⛔ Hang up immediately."
            : "You are on an active call.\n\nNever share your OTP verbally.\nOnly enter it directly on the app screen."
        stack.addArrangedSubview(body)

        if riskLevel == .highRisk {
            let btn = UIButton(type: .system)
            btn.setTitle("I understand — End Call Now", for: .normal)
            btn.backgroundColor = UIColor.white
            btn.setTitleColor(UIColor(red: 0.78, green: 0.10, blue: 0.10, alpha: 1), for: .normal)
            btn.titleLabel?.font = .boldSystemFont(ofSize: 15)
            btn.layer.cornerRadius = 10
            btn.contentEdgeInsets = UIEdgeInsets(top: 12, left: 24, bottom: 12, right: 24)
            btn.addTarget(self, action: #selector(dismiss(_:)), for: .touchUpInside)
            stack.addArrangedSubview(btn)
        }
    }

    @objc private func dismiss(_ sender: Any) {
        dismiss(animated: true)
    }
}

File: sdk/android/src/main/java/com/swifter/fraudshield/OtpGuard.kt
/* ═══════════════════════════════════════════════════════════════════════════
 * PayGuard Android SDK — OTP Guard
 * Protects OTP screens from social engineering disclosure
 * © 2026 Swifter Technologies (Pty) Ltd
 * ═══════════════════════════════════════════════════════════════════════════ */

package com.swifter.fraudshield

import android.app.Activity
import android.app.AlertDialog
import android.graphics.Color
import android.graphics.drawable.ColorDrawable
import android.view.Gravity
import android.view.ViewGroup.LayoutParams
import android.view.WindowManager
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import com.swifter.fraudshield.collectors.CallStateCollector
import com.swifter.fraudshield.network.FraudSignalDispatcher
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * OTP Guard — Prevents OTP vishing / social engineering disclosure.
 *
 * When a bank's OTP screen opens while the user is on a phone call,
 * this module:
 *
 *   1. Applies FLAG_SECURE to prevent screenshots and screen recording
 *   2. Shows a full-screen red warning overlay:
 *      - Unknown caller → "SCAM ALERT" with instructions to hang up
 *      - Known caller → gentler warning about not sharing OTP verbally
 *   3. Fires RULE_014 signal (OTP_SCREEN_ON_CALL) to the risk engine
 *      with risk delta 80 (unknown) or 45 (known)
 *
 * INTEGRATION IN BANK APP:
 *
 *   class OtpActivity : AppCompatActivity() {
 *       private val otpGuard = OtpGuard.shared
 *
 *       override fun onResume() {
 *           super.onResume()
 *           otpGuard.activate(this) { riskLevel ->
 *               if (riskLevel == OtpRiskLevel.HIGH_RISK) {
 *                   // Optionally auto-cancel the OTP
 *               }
 *           }
 *       }
 *
 *       override fun onPause() {
 *           super.onPause()
 *           otpGuard.deactivate(this)
 *       }
 *   }
 */

enum class OtpRiskLevel {
    CLEAN,        // Not on a call
    MEDIUM_RISK,  // On a known call — warning shown
    HIGH_RISK     // On a call with unknown caller — full SCAM ALERT
}

class OtpGuard private constructor() {

    companion object {
        val shared = OtpGuard()
    }

    private var warningDialog: AlertDialog? = null
    private var callCollector: CallStateCollector? = null

    // ── Public API ──────────────────────────────────────────────────────────

    /**
     * Activate OTP protection on the given Activity.
     *
     * Checks if the user is on an active call. If so, applies FLAG_SECURE
     * and shows a warning overlay. Fires the OTP_SCREEN_ON_CALL signal.
     *
     * @param activity The OTP screen activity
     * @param callback Reports the detected risk level back to the host app
     */
    fun activate(
        activity: Activity,
        callback: (OtpRiskLevel) -> Unit
    ) {
        // Step 1: Apply FLAG_SECURE — prevents screenshots/screen recording
        activity.window.setFlags(
            WindowManager.LayoutParams.FLAG_SECURE,
            WindowManager.LayoutParams.FLAG_SECURE
        )

        // Step 2: Check call state
        if (callCollector == null) {
            callCollector = CallStateCollector(activity.applicationContext)
        }

        if (!callCollector!!.isOnCall()) {
            callback(OtpRiskLevel.CLEAN)
            return
        }

        // User IS on a call — determine risk level
        val isUnknownCaller = callCollector!!.isCallerUnknown()
        val riskLevel = if (isUnknownCaller) OtpRiskLevel.HIGH_RISK else OtpRiskLevel.MEDIUM_RISK

        // Step 3: Show warning overlay
        showWarningOverlay(activity, riskLevel)

        // Step 4: Fire risk signal
        fireOtpOnCallSignal(riskLevel)

        callback(riskLevel)
    }

    /**
     * Deactivate OTP protection. Call from onPause().
     * Removes FLAG_SECURE and dismisses the warning overlay.
     */
    fun deactivate(activity: Activity) {
        activity.window.clearFlags(WindowManager.LayoutParams.FLAG_SECURE)
        dismissWarning()
    }

    // ── Warning Overlay ─────────────────────────────────────────────────────

    private fun showWarningOverlay(activity: Activity, riskLevel: OtpRiskLevel) {
        dismissWarning()

        val isHighRisk = riskLevel == OtpRiskLevel.HIGH_RISK
        val dangerRed = Color.parseColor("#C81919")

        val layout = LinearLayout(activity).apply {
            orientation = LinearLayout.VERTICAL
            gravity = Gravity.CENTER
            setPadding(64, 96, 64, 96)
            setBackgroundColor(dangerRed)
        }

        // Emoji
        val emoji = TextView(activity).apply {
            text = if (isHighRisk) "🚨" else "⚠️"
            textSize = 56f
            gravity = Gravity.CENTER
        }
        layout.addView(emoji)

        // Headline
        val headline = TextView(activity).apply {
            text = if (isHighRisk) "SCAM ALERT" else "Active Call Detected"
            textSize = 22f
            setTextColor(Color.WHITE)
            gravity = Gravity.CENTER
            setPadding(0, 24, 0, 16)
        }
        layout.addView(headline)

        // Body
        val body = TextView(activity).apply {
            text = if (isHighRisk) {
                "You are on a call with an UNKNOWN number.\n\n" +
                "⛔ NEVER read your OTP to anyone.\n" +
                "⛔ Your bank will never ask for your code.\n" +
                "⛔ Hang up immediately."
            } else {
                "You are on an active call.\n\n" +
                "Never share your OTP verbally.\n" +
                "Only enter it directly on the app screen."
            }
            textSize = 15f
            setTextColor(Color.parseColor("#E5E5E5"))
            gravity = Gravity.CENTER
            setPadding(0, 0, 0, 32)
        }
        layout.addView(body)

        // Dismiss button (high risk only)
        if (isHighRisk) {
            val button = Button(activity).apply {
                text = "I understand — End Call Now"
                textSize = 15f
                setTextColor(dangerRed)
                setBackgroundColor(Color.WHITE)
                setPadding(48, 24, 48, 24)
                setOnClickListener { dismissWarning() }
            }
            layout.addView(button, LinearLayout.LayoutParams(
                LayoutParams.WRAP_CONTENT,
                LayoutParams.WRAP_CONTENT
            ).apply { gravity = Gravity.CENTER })
        }

        warningDialog = AlertDialog.Builder(activity, android.R.style.Theme_NoTitleBar_Fullscreen)
            .setView(layout)
            .setCancelable(!isHighRisk) // High risk: must tap button to dismiss
            .create()
            .also { dialog ->
                dialog.window?.setBackgroundDrawable(ColorDrawable(dangerRed))
                dialog.show()
            }
    }

    private fun dismissWarning() {
        warningDialog?.dismiss()
        warningDialog = null
    }

    // ── Risk Signal Dispatch ────────────────────────────────────────────────

    private fun fireOtpOnCallSignal(riskLevel: OtpRiskLevel) {
        val dispatcher = FraudSignalDispatcher.shared ?: return

        val event = mapOf(
            "event_type" to "OTP_SCREEN_ON_CALL",
            "unknown_caller" to (riskLevel == OtpRiskLevel.HIGH_RISK),
            "rule" to "RULE_014",
            "risk_delta" to if (riskLevel == OtpRiskLevel.HIGH_RISK) 80 else 45,
            "timestamp" to (System.currentTimeMillis() / 1000.0)
        )

        CoroutineScope(Dispatchers.IO).launch {
            try {
                dispatcher.dispatchEvent(event)
            } catch (_: Exception) {
                // Non-blocking
            }
        }
    }
}
